Applications/Spire/Source/Ui/ScrollArea.cpp

```diff
@@ -0,0 +1,156 @@
+  verticalScrollBar()->setContextMenuPolicy(Qt::NoContextMenu);
+  setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+  setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+  m_horizontal_scroll_bar_timer.setInterval(SCROLL_BAR_HIDE_TIME_MS);
+  connect(&m_horizontal_scroll_bar_timer, &QTimer::timeout, this,
+    &ScrollArea::hide_horizontal_scroll_bar);
+  m_vertical_scroll_bar_timer.setInterval(SCROLL_BAR_HIDE_TIME_MS);
+  connect(&m_vertical_scroll_bar_timer, &QTimer::timeout, this,
+    &ScrollArea::hide_vertical_scroll_bar);
+}
+
+void ScrollArea::setWidget(QWidget* widget) {
+  widget->setMouseTracking(true);
+  widget->setAttribute(Qt::WA_Hover);
+  widget->installEventFilter(this);
+  QScrollArea::setWidget(widget);
+}
+
+bool ScrollArea::eventFilter(QObject* watched, QEvent* event) {
+  if(event->type() == QEvent::HoverMove) {
+    auto e = static_cast<QHoverEvent*>(event);
+    if(is_within_opposite_scroll_bar(verticalScrollBar(), e->pos().y(),
+        widget()->height(), height()) && !verticalScrollBar()->isVisible()) {
+      set_scroll_bar_style(SCROLL_BAR_MAX_SIZE);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else if(is_within_opposite_scroll_bar(horizontalScrollBar(),
+        e->pos().x(), widget()->width(), width()) &&
+        !horizontalScrollBar()->isSliderDown()) {
+      set_scroll_bar_style(SCROLL_BAR_MAX_SIZE);
+      setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
+      setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    } else {
+      set_scroll_bar_style(SCROLL_BAR_MIN_SIZE);
+      if(!m_vertical_scroll_bar_timer.isActive() &&
+          verticalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !verticalScrollBar()->isSliderDown()) {
+        hide_vertical_scroll_bar();
+      }
+      if(!m_horizontal_scroll_bar_timer.isActive() &&
+          horizontalScrollBarPolicy() != Qt::ScrollBarAlwaysOff &&
+          !horizontalScrollBar()->isSliderDown()) {
+        hide_horizontal_scroll_bar();
+      }
+    }
+  } else if(event->type() == QEvent::HoverLeave) {
+    if(!m_horizontal_scroll_bar_timer.isActive() &&
+        horizontalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_horizontal_scroll_bar();
+    }
+    if(!m_vertical_scroll_bar_timer.isActive() &&
+        verticalScrollBarPolicy() == Qt::ScrollBarAlwaysOff) {
+      hide_vertical_scroll_bar();
+    }
+  }
+  return QScrollArea::eventFilter(watched, event);
+}
+
+void ScrollArea::leaveEvent(QEvent* event) {
+  hide_horizontal_scroll_bar();
+  hide_vertical_scroll_bar();
+}
+
+void ScrollArea::wheelEvent(QWheelEvent* event) {
+  if(event->modifiers().testFlag(Qt::ShiftModifier)) {
+    setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
+    setHorizontalScrollBarPolicy(Qt::ScrollBarAsNeeded);
+    horizontalScrollBar()->setValue(horizontalScrollBar()->value() -
+      (event->delta() / 2));
```

Comment:
I am worried about the accumulation of errors here. If the delta is an odd number then this is off by 1. Imagine if the delta is always 1, then no change ever occurs.

Reply:
It looks like delta() was actually deprecated. I can use the [angle delta](https://doc.qt.io/qt-5/qwheelevent.html#angleDelta) instead.

What about taking the larger of (angleDelta / 2) or 1?

Reply:
You can use angle delta, that's fine, but it doesn't solve the accumulation of errors. To solve that you will need to store the accumulated error and add it back in. Otherwise you will need to find some objective value to use to update the scrollbar's value.

Reply:
Or -1, if the angleDelta is negative.

Reply:
Keep track of the rounding error and add it back in when it's equal to 1.

Reply:
Would it look something like this:

`
auto delta = angleDelta() / 2 + accumulated_error;
accumulated_error = modf(delta, &delta);
scrollBar()->setValue(delta);
`

Reply:
Sorry, I keep missing your replies, the page isn't refreshing to show them.

Reply:
Yeah something like that is about right. accumulated error should just be either 0 or 0.5.